File: dashboard/lib/overview_query.py
#!/usr/bin/env python3
"""Compute dashboard overview metrics from MongoDB."""

from __future__ import annotations

import argparse
import json
import math
import os
from collections import defaultdict
from datetime import datetime, timedelta, timezone
from typing import Any

from pymongo import MongoClient

ENDPOINT_FAMILY_CODING_PLAN = "coding_plan"
ENDPOINT_FAMILY_OFFICIAL_API = "official_api"
ENDPOINT_FAMILIES = (ENDPOINT_FAMILY_CODING_PLAN, ENDPOINT_FAMILY_OFFICIAL_API)
MIN_STABLE_GENERATION_WINDOW_MS = 500.0
MAX_REASONABLE_TPS = 1000.0


def _percentile(values: list[float], p: float) -> float | None:
    if not values:
        return None
    ordered = sorted(values)
    if len(ordered) == 1:
        return ordered[0]
    rank = (len(ordered) - 1) * p
    lower = math.floor(rank)
    upper = math.ceil(rank)
    if lower == upper:
        return ordered[int(rank)]
    weight = rank - lower
    return ordered[lower] * (1 - weight) + ordered[upper] * weight


def _to_iso(value: datetime | None) -> str | None:
    if value is None:
        return None
    if value.tzinfo is None:
        value = value.replace(tzinfo=timezone.utc)
    return value.isoformat()


def _as_utc(value: datetime) -> datetime:
    if value.tzinfo is None:
        return value.replace(tzinfo=timezone.utc)
    return value.astimezone(timezone.utc)


def _next_thirty_mark(now: datetime) -> datetime:
    now = now.astimezone(timezone.utc)
    if now.minute < 30:
        return now.replace(minute=30, second=0, microsecond=0)
    return (now + timedelta(hours=1)).replace(minute=30, second=0, microsecond=0)


def _normalize_endpoint_family(raw: str) -> str:
    normalized = raw.strip().lower().replace("-", "_")
    if normalized in ENDPOINT_FAMILIES:
        return normalized
    raise ValueError(f"endpoint family must be one of: {', '.join(ENDPOINT_FAMILIES)}")


def _build_endpoint_family_match(endpoint_family: str) -> dict[str, Any]:
    if endpoint_family == ENDPOINT_FAMILY_CODING_PLAN:
        return {
            "$or": [
                {"endpoint_family": ENDPOINT_FAMILY_CODING_PLAN},
                {
                    "$and": [
                        {"endpoint_family": {"$exists": False}},
                        {"endpoint_base": {"$regex": r"/api/coding/paas/v4/?$", "$options": "i"}},
                    ]
                },
            ]
        }

    return {
        "$or": [
            {"endpoint_family": ENDPOINT_FAMILY_OFFICIAL_API},
            {
                "$and": [
                    {"endpoint_family": {"$exists": False}},
                    {"endpoint_base": {"$regex": r"/api/paas/v4/?$", "$options": "i"}},
                ]
            },
        ]
    }


def _extract_stable_tps(doc: dict[str, Any], key: str) -> float | None:
    metrics = doc.get("metrics", {})
    if not isinstance(metrics, dict):
        return None

    raw_value = metrics.get(key)
    raw_generation_window_ms = metrics.get("generation_window_ms")
    if raw_value is None or raw_generation_window_ms is None:
        return None

    try:
        generation_window_ms = float(raw_generation_window_ms)
    except (TypeError, ValueError):
        return None
    if generation_window_ms < MIN_STABLE_GENERATION_WINDOW_MS:
        return None

    try:
        value = float(raw_value)
    except (TypeError, ValueError):
        return None
    if value < 0 or value > MAX_REASONABLE_TPS:
        return None
    return value


def _extract_output_tps_post_ttft(doc: dict[str, Any]) -> float | None:
    metrics = doc.get("metrics", {})
    if not isinstance(metrics, dict):
        return None

    raw_stored = metrics.get("output_tokens_per_second_post_ttft")
    if raw_stored is not None:
        try:
            stored_value = float(raw_stored)
        except (TypeError, ValueError):
            stored_value = None
        if stored_value is not None and stored_value >= 0:
            return stored_value

    tokens = doc.get("tokens", {})
    if not isinstance(tokens, dict):
        return None

    raw_completion_tokens = tokens.get("completion_tokens")
    raw_total_latency_ms = metrics.get("total_latency_ms")
    raw_ttft_ms = metrics.get("ttft_ms")
    if raw_completion_tokens is None or raw_total_latency_ms is None or raw_ttft_ms is None:
        return None

    try:
        completion_tokens = float(raw_completion_tokens)
        total_latency_ms = float(raw_total_latency_ms)
        ttft_ms = float(raw_ttft_ms)
    except (TypeError, ValueError):
        return None

    if completion_tokens <= 1 or total_latency_ms <= ttft_ms:
        return None

    return (completion_tokens - 1) / ((total_latency_ms - ttft_ms) / 1000)


def _as_dict(value: Any) -> dict[str, Any]:
    return value if isinstance(value, dict) else {}


def _collect_metric_values(docs: list[dict[str, Any]], key: str) -> list[float]:
    values: list[float] = []
    for doc in docs:
        raw = _as_dict(doc.get("metrics")).get(key)
        if raw is None:
            continue
        try:
            values.append(float(raw))
        except (TypeError, ValueError):
            continue
    return values


def _collect_metric_gap_values(docs: list[dict[str, Any]], left_key: str, right_key: str) -> list[float]:
    values: list[float] = []
    for doc in docs:
        metrics = _as_dict(doc.get("metrics"))
        left_raw = metrics.get(left_key)
        right_raw = metrics.get(right_key)
        if left_raw is None or right_raw is None:
            continue
        try:
            values.append(float(left_raw) - float(right_raw))
        except (TypeError, ValueError):
            continue
    return values


def _parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--hours", type=float, default=24.0)
    parser.add_argument("--model", default="")
    parser.add_argument("--endpoint-family", default=ENDPOINT_FAMILY_CODING_PLAN)
    return parser.parse_args()


def main() -> int:
    args = _parse_args()
    try:
        endpoint_family = _normalize_endpoint_family(args.endpoint_family)
    except ValueError as exc:
        print(json.dumps({"error": str(exc)}))
        return 2

    mongodb_uri = os.getenv("MONGODB_URI")
    if not mongodb_uri:
        print(json.dumps({"error": "MONGODB_URI is required"}))
        return 2

    db_name = os.getenv("MONGO_DB", "zaimonitor")
    collection_name = os.getenv("MONGO_COLLECTION", "inference_runs")

    requested_hours = max(args.hours, 24.0)
    scope_days = max(1, math.ceil(requested_hours / 24.0))
    now_utc = datetime.now(timezone.utc)
    current_midnight_utc = now_utc.replace(hour=0, minute=0, second=0, microsecond=0)
    metrics_window_start = now_utc - timedelta(hours=24)
    metrics_window_end = now_utc

    client = MongoClient(mongodb_uri, serverSelectionTimeoutMS=10000)
    collection = client[db_name][collection_name]

    scope_filter = _build_endpoint_family_match(endpoint_family)
    trend_scope_filter: dict[str, Any] = {**scope_filter}
    if args.model:
        trend_scope_filter["model"] = args.model

    latest_trend_doc = next(
        iter(
            collection.find(
                trend_scope_filter,
                projection={"_id": 0, "timestamp": 1},
            )
            .sort("timestamp", -1)
            .limit(1)
        ),
        None,
    )
    latest_trend_timestamp = (
        latest_trend_doc.get("timestamp")
        if isinstance(latest_trend_doc, dict)
        else None
    )
    if isinstance(latest_trend_timestamp, datetime):
        latest_day_start = _as_utc(latest_trend_timestamp).replace(hour=0, minute=0, second=0, microsecond=0)
        trend_window_end = latest_day_start + timedelta(days=1)
    else:
        trend_window_end = current_midnight_utc
    trend_window_start = trend_window_end - timedelta(days=scope_days)

    match_metrics: dict[str, Any] = {
        "timestamp": {"$gte": metrics_window_start, "$lt": metrics_window_end},
        **scope_filter,
    }
    match_trend: dict[str, Any] = {
        "timestamp": {"$gte": trend_window_start, "$lt": trend_window_end},
        **scope_filter,
    }
    # Model options should not disappear when time/model filters narrow the main dataset.
    match_models: dict[str, Any] = {
        **scope_filter,
    }
    if args.model:
        match_metrics["model"] = args.model
        match_trend["model"] = args.model

    projection = {
        "_id": 0,
        "timestamp": 1,
        "metrics_version": 1,
        "ok": 1,
        "model": 1,
        "endpoint_family": 1,
        "endpoint_base": 1,
        "metrics.first_sse_event_ms": 1,
        "metrics.first_reasoning_token_ms": 1,
        "metrics.first_answer_token_ms": 1,
        "metrics.ttft_ms": 1,
        "metrics.thinking_window_ms": 1,
        "metrics.time_to_completed_answer_ms": 1,
        "metrics.provider_output_tokens_per_second": 1,
        "metrics.provider_output_tokens_per_second_end_to_end": 1,
        "metrics.output_tokens_per_second_post_ttft": 1,
        "metrics.visible_output_tokens_per_second": 1,
        "metrics.generation_window_ms": 1,
        "metrics.total_latency_ms": 1,
        "tokens.completion_tokens": 1,
        "error.type": 1,
    }

    docs_v4 = list(
        collection.find(
            {**match_metrics, "metrics_version": {"$gte": 4}},
            projection=projection,
        ).sort("timestamp", 1)
    )
    using_legacy_metrics = False
    docs = docs_v4
    if not docs:
        using_legacy_metrics = True
        docs = list(collection.find(match_metrics, projection=projection).sort("timestamp", 1))

    trend_docs_v4 = list(
        collection.find(
            {**match_trend, "metrics_version": {"$gte": 4}},
            projection=projection,
        ).sort("timestamp", 1)
    )
    trend_docs = trend_docs_v4
    if not trend_docs:
        trend_docs = list(collection.find(match_trend, projection=projection).sort("timestamp", 1))

    total_requests = len(docs)
    success_docs = [d for d in docs if d.get("ok")]
    failure_docs = [d for d in docs if not d.get("ok")]
    trend_success_docs = [d for d in trend_docs if d.get("ok")]

    ttft_values = _collect_metric_values(success_docs, "ttft_ms")
    first_sse_values = _collect_metric_values(success_docs, "first_sse_event_ms")
    first_reasoning_values = _collect_metric_values(success_docs, "first_reasoning_token_ms")
    first_answer_values = _collect_metric_values(success_docs, "first_answer_token_ms")
    sse_to_visible_gap_values = _collect_metric_gap_values(success_docs, "ttft_ms", "first_sse_event_ms")
    thinking_window_values = _collect_metric_values(success_docs, "thinking_window_ms")
    completed_answer_values = _collect_metric_values(success_docs, "time_to_completed_answer_ms")
    provider_tps_values = [
        value
        for d in success_docs
        for value in [_extract_stable_tps(d, "provider_output_tokens_per_second")]
        if value is not None
    ]
    provider_tps_e2e_values = _collect_metric_values(success_docs, "provider_output_tokens_per_second_end_to_end")
    output_tps_values = [
        value
        for d in success_docs
        for value in [_extract_output_tps_post_ttft(d)]
        if value is not None
    ]
    visible_tps_values = [
        value
        for d in success_docs
        for value in [_extract_stable_tps(d, "visible_output_tokens_per_second")]
        if value is not None
    ]
    total_latency_values = _collect_metric_values(success_docs, "total_latency_ms")

    buckets: dict[str, dict[str, float]] = defaultdict(
        lambda: {
            "output_sum": 0.0,
            "output_count": 0.0,
            "visible_sum": 0.0,
            "visible_count": 0.0,
            "provider_sum": 0.0,
            "provider_count": 0.0,
        }
    )
    for d in trend_success_docs:
        output_tps = _extract_output_tps_post_ttft(d)
        visible_tps = _extract_stable_tps(d, "visible_output_tokens_per_second")
        provider_tps = _extract_stable_tps(d, "provider_output_tokens_per_second")
        ts = d.get("timestamp")
        if not isinstance(ts, datetime):
            continue
        bucket = ts.astimezone(timezone.utc).replace(minute=0, second=0, microsecond=0)
        key = bucket.isoformat()
        if output_tps is not None:
            buckets[key]["output_sum"] += float(output_tps)
            buckets[key]["output_count"] += 1
        if visible_tps is not None:
            buckets[key]["visible_sum"] += float(visible_tps)
            buckets[key]["visible_count"] += 1
        if provider_tps is not None:
            buckets[key]["provider_sum"] += float(provider_tps)
            buckets[key]["provider_count"] += 1

    trend = []
    bucket_cursor = trend_window_start
    while bucket_cursor < trend_window_end:
        key = bucket_cursor.isoformat()
        data = buckets.get(key)
        trend.append(
            {
                "timestamp": key,
                "output_tps": round(data["output_sum"] / data["output_count"], 3)
                if data and data["output_count"] > 0
                else None,
                "visible_tps": round(data["visible_sum"] / data["visible_count"], 3)
                if data and data["visible_count"] > 0
                else None,
                "provider_tps": round(data["provider_sum"] / data["provider_count"], 3)
                if data and data["provider_count"] > 0
                else None,
            }
        )
        bucket_cursor += timedelta(hours=1)

    error_breakdown: dict[str, int] = defaultdict(int)
    for d in failure_docs:
        error_type = _as_dict(d.get("error")).get("type") or "unknown_error"
        error_breakdown[error_type] += 1

    models = sorted(
        [
            model
            for model in collection.distinct("model", match_models)
            if isinstance(model, str) and model
        ]
    )
    if not models:
        models = ["glm-4.7", "glm-5"]
    latest_ts = docs[-1].get("timestamp") if docs else (trend_docs[-1].get("timestamp") if trend_docs else None)

    next_run = _next_thirty_mark(now_utc)
    cadence_label = "Updates every hour"

    payload = {
        "window": {
            "hours": scope_days * 24,
            "start": _to_iso(trend_window_start),
            "end": _to_iso(trend_window_end),
        },
        "totals": {
            "requests": total_requests,
            "successes": len(success_docs),
            "failures": len(failure_docs),
            "success_rate_percent": round((len(success_docs) / total_requests) * 100, 2)
            if total_requests
            else None,
        },
        "metrics": {
            "avg_first_sse_event_ms": round(sum(first_sse_values) / len(first_sse_values), 2)
            if first_sse_values
            else None,
            "avg_first_reasoning_token_ms": round(sum(first_reasoning_values) / len(first_reasoning_values), 2)
            if first_reasoning_values
            else None,
            "avg_first_answer_token_ms": round(sum(first_answer_values) / len(first_answer_values), 2)
            if first_answer_values
            else None,
            "avg_ttft_ms": round(sum(ttft_values) / len(ttft_values), 2) if ttft_values else None,
            "avg_sse_to_visible_gap_ms": round(sum(sse_to_visible_gap_values) / len(sse_to_visible_gap_values), 2)
            if sse_to_visible_gap_values
            else None,
            "avg_thinking_window_ms": round(sum(thinking_window_values) / len(thinking_window_values), 2)
            if thinking_window_values
            else None,
            "avg_time_to_completed_answer_ms": round(sum(completed_answer_values) / len(completed_answer_values), 2)
            if completed_answer_values
            else None,
            "avg_visible_tps": round(sum(visible_tps_values) / len(visible_tps_values), 3)
            if visible_tps_values
            else None,
            "avg_output_tps": round(sum(output_tps_values) / len(output_tps_values), 3)
            if output_tps_values
            else None,
            "avg_provider_tps": round(sum(provider_tps_values) / len(provider_tps_values), 3)
            if provider_tps_values
            else None,
            "avg_provider_tps_end_to_end": round(sum(provider_tps_e2e_values) / len(provider_tps_e2e_values), 3)
            if provider_tps_e2e_values
            else None,
            "p95_total_latency_ms": round(_percentile(total_latency_values, 0.95), 2)
            if total_latency_values
            else None,
        },
        "trend": trend,
        "errors": [{"type": k, "count": v} for k, v in sorted(error_breakdown.items(), key=lambda kv: kv[1], reverse=True)],
        "models": models,
        "endpoint_families": list(ENDPOINT_FAMILIES),
        "selected_endpoint_family": endpoint_family,
        "selected_model": args.model or None,
        "using_legacy_metrics": using_legacy_metrics,
        "latest_document_timestamp": _to_iso(latest_ts if isinstance(latest_ts, datetime) else None),
        "schedule": {
            "cadence_label": cadence_label,
            "next_run_utc": _to_iso(next_run),
        },
        "generated_at": _to_iso(datetime.now(timezone.utc)),
    }

    print(json.dumps(payload, ensure_ascii=False))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: dashboard/lib/overview-query.ts
import { MongoClient, Collection, Document } from "mongodb";
import { ALL_MODELS } from "./constants";

const ENDPOINT_FAMILY_CODING_PLAN = "coding_plan";

interface Metrics {
  ttft_ms?: number;
  provider_output_tokens_per_second_end_to_end?: number;
  output_tokens_per_second_post_ttft?: number;
  total_latency_ms?: number;
}

interface Tokens {
  completion_tokens?: number;
}

interface ErrorDoc {
  type?: string;
}

interface InferenceDoc extends Document {
  timestamp: Date;
  run_id?: string;
  ok: boolean;
  model?: string;
  metrics?: Metrics;
  tokens?: Tokens;
  error?: ErrorDoc;
}

interface BucketData {
  output_sum: number;
  output_count: number;
  ttft_sum: number;
  ttft_count: number;
}

interface RunTrendPoint {
  timestamp: Date;
  model: string;
  output_tps?: number;
  ttft_ms?: number;
}

type MongoClientCache = {
  clients: Map<string, Promise<MongoClient>>;
};

declare global {
  var __zaiMongoClientCache: MongoClientCache | undefined;
}

function getMongoClientCache(): MongoClientCache {
  if (!globalThis.__zaiMongoClientCache) {
    globalThis.__zaiMongoClientCache = { clients: new Map<string, Promise<MongoClient>>() };
  }
  return globalThis.__zaiMongoClientCache;
}

async function getMongoClient(mongoUri: string): Promise<MongoClient> {
  const cache = getMongoClientCache();
  const existing = cache.clients.get(mongoUri);
  if (existing) return existing;

  const clientPromise = new MongoClient(mongoUri, { serverSelectionTimeoutMS: 10000 })
    .connect()
    .catch((error) => {
      cache.clients.delete(mongoUri);
      throw error;
    });

  cache.clients.set(mongoUri, clientPromise);
  return clientPromise;
}

function percentile(values: number[], p: number): number | null {
  if (!values.length) return null;
  const ordered = [...values].sort((a, b) => a - b);
  if (ordered.length === 1) return ordered[0];
  const rank = (ordered.length - 1) * p;
  const lower = Math.floor(rank);
  const upper = Math.ceil(rank);
  if (lower === upper) return ordered[Math.floor(rank)];
  const weight = rank - lower;
  return ordered[lower] * (1 - weight) + ordered[upper] * weight;
}

function toIso(value: Date | null | undefined): string | null {
  if (!value) return null;
  if (!(value instanceof Date)) return null;
  const date = new Date(value);
  if (isNaN(date.getTime())) return null;
  return date.toISOString();
}

function buildEndpointFamilyMatch(): Document {
  return {
    $or: [
      { endpoint_family: ENDPOINT_FAMILY_CODING_PLAN },
      {
        $and: [
          { endpoint_family: { $exists: false } },
          { endpoint_base: { $regex: /\/api\/coding\/paas\/v4\/?$/i } },
        ],
      },
    ],
  };
}

function extractOutputTpsPostTtft(doc: InferenceDoc): number | null {
  const metrics = doc.metrics || {};
  
  const storedValue = metrics.output_tokens_per_second_post_ttft;
  if (storedValue !== undefined) {
    const value = Number(storedValue);
    if (!isNaN(value) && value >= 0) return value;
  }

  const tokens = doc.tokens || {};
  const completionTokens = tokens.completion_tokens;
  const totalLatencyMs = metrics.total_latency_ms;
  const ttftMs = metrics.ttft_ms;

  if (completionTokens === undefined || totalLatencyMs === undefined || ttftMs === undefined) return null;

  const ct = Number(completionTokens);
  const tl = Number(totalLatencyMs);
  const ttft = Number(ttftMs);

  if (isNaN(ct) || isNaN(tl) || isNaN(ttft)) return null;
  if (ct <= 1 || tl <= ttft) return null;

  return (ct - 1) / ((tl - ttft) / 1000);
}

function normalizeModel(model: string | undefined): string {
  if (!model || typeof model !== "string") return "unknown";
  const trimmed = model.trim().toLowerCase();
  if (
    trimmed.includes("glm-4.7-flash")
    || trimmed.includes("glm47-flash")
    || trimmed.includes("glm47flash")
  ) {
    return "glm-4.7-flash";
  }
  if (trimmed.includes("glm-5") || trimmed === "glm5") return "glm-5";
  if (trimmed.includes("glm-4.7") || trimmed.includes("glm47")) return "glm-4.7";
  return trimmed || "unknown";
}

function initBucketData(): BucketData {
  return { output_sum: 0, output_count: 0, ttft_sum: 0, ttft_count: 0 };
}

function buildRunTrendPoints(docs: InferenceDoc[]): RunTrendPoint[] {
  const runBuckets = new Map<string, { timestamp: Date; model: string; data: BucketData }>();

  for (const [index, doc] of docs.entries()) {
    const ts = doc.timestamp;
    if (!(ts instanceof Date) || isNaN(ts.getTime())) continue;

    const rawRunId = typeof doc.run_id === "string" ? doc.run_id.trim() : "";
    const model = normalizeModel(doc.model);
    const runKey = rawRunId.length > 0 ? `${rawRunId}|${model}` : `legacy:${index}|${model}`;
    const existing = runBuckets.get(runKey) || { timestamp: new Date(ts), model, data: initBucketData() };

    if (ts.getTime() > existing.timestamp.getTime()) {
      existing.timestamp = new Date(ts);
    }

    if (doc.ok) {
      const outputTps = extractOutputTpsPostTtft(doc);
      if (outputTps !== null) {
        existing.data.output_sum += outputTps;
        existing.data.output_count += 1;
      }

      const ttftRaw = doc.metrics?.ttft_ms;
      const ttft = ttftRaw == null ? null : Number(ttftRaw);
      if (ttft !== null && !isNaN(ttft)) {
        existing.data.ttft_sum += ttft;
        existing.data.ttft_count += 1;
      }
    }

    runBuckets.set(runKey, existing);
  }

  const runTrendPoints: RunTrendPoint[] = [];
  for (const { timestamp, model, data } of runBuckets.values()) {
    runTrendPoints.push({
      timestamp,
      model,
      output_tps: data.output_count > 0 ? data.output_sum / data.output_count : undefined,
      ttft_ms: data.ttft_count > 0 ? data.ttft_sum / data.ttft_count : undefined,
    });
  }

  runTrendPoints.sort((a, b) => a.timestamp.getTime() - b.timestamp.getTime());
  return runTrendPoints;
}

function collectMetricValues(docs: InferenceDoc[], key: keyof Metrics): number[] {
  const values: number[] = [];
  for (const doc of docs) {
    const raw = doc.metrics?.[key];
    if (raw == null) continue;
    const value = Number(raw);
    if (!isNaN(value)) values.push(value);
  }
  return values;
}

export interface OverviewQueryParams {
  hours: number;
}

export interface ModelMetrics {
  requests: number;
  successes: number;
  failures: number;
  success_rate_percent: number | null;
  avg_ttft_ms: number | null;
  avg_output_tps: number | null;
  p95_ttft_ms: number | null;
  avg_provider_tps_end_to_end: number | null;
}

export interface OverviewResult {
  window: {
    hours: number;
    start: string | null;
    end: string | null;
  };
  metrics_by_model: Record<string, ModelMetrics>;
  trend_by_model: Record<string, Array<{
    timestamp: string;
    output_tps?: number;
    ttft_ms?: number;
  }>>;
  failure_by_model: Record<string, Array<{ timestamp: string }>>;
  errors: Array<{ type: string; count: number }>;
  models: string[];
  endpoint_family: typeof ENDPOINT_FAMILY_CODING_PLAN;
  latest_document_timestamp: string | null;
  generated_at: string | null;
}

export async function queryOverview(
  mongoUri: string,
  params: OverviewQueryParams,
): Promise<OverviewResult> {
  const dbName = process.env.MONGO_DB || "zaimonitor";
  const collectionName = process.env.MONGO_COLLECTION || "inference_runs";

  const requestedHours = Math.max(params.hours, 24);
  const trendWindowDurationMs = requestedHours * 60 * 60 * 1000;
  const nowUtc = new Date();
  const metricsWindowStart = new Date(nowUtc.getTime() - requestedHours * 60 * 60 * 1000);

  const client = await getMongoClient(mongoUri);
  const db = client.db(dbName);
  const collection: Collection<InferenceDoc> = db.collection(collectionName);

  const scopeFilter = buildEndpointFamilyMatch();

  const latestDoc = await collection
    .findOne(scopeFilter, { projection: { _id: 0, timestamp: 1 }, sort: { timestamp: -1 } })
    .catch(() => null);

  const latestTimestamp = latestDoc?.timestamp;
  const trendWindowEnd =
    latestTimestamp instanceof Date && !isNaN(latestTimestamp.getTime())
      ? new Date(latestTimestamp)
      : nowUtc;
  const trendWindowStart = new Date(trendWindowEnd.getTime() - trendWindowDurationMs);

  const matchMetrics: Document = {
    timestamp: { $gte: metricsWindowStart, $lt: nowUtc },
    ...scopeFilter,
  };
  const matchTrend: Document = {
    timestamp: { $gte: trendWindowStart, $lte: trendWindowEnd },
    ...scopeFilter,
  };

  const projection = {
    _id: 0,
    timestamp: 1,
    run_id: 1,
    ok: 1,
    model: 1,
    "metrics.ttft_ms": 1,
    "metrics.provider_output_tokens_per_second_end_to_end": 1,
    "metrics.output_tokens_per_second_post_ttft": 1,
    "metrics.total_latency_ms": 1,
    "tokens.completion_tokens": 1,
    "error.type": 1,
  };

  const [docsV4, trendDocsV4] = await Promise.all([
    collection.find({ ...matchMetrics, metrics_version: { $gte: 4 } }, { projection }).sort({ timestamp: 1 }).toArray(),
    collection.find({ ...matchTrend, metrics_version: { $gte: 4 } }, { projection }).sort({ timestamp: 1 }).toArray(),
  ]);

  let docs = docsV4;
  let trendDocs = trendDocsV4;
  if (!docs.length) {
    docs = await collection.find(matchMetrics, { projection }).sort({ timestamp: 1 }).toArray();
  }
  if (!trendDocs.length) {
    trendDocs = await collection.find(matchTrend, { projection }).sort({ timestamp: 1 }).toArray();
  }

  const failureDocs = docs.filter((d) => !d.ok);
  const runTrendPoints = buildRunTrendPoints(trendDocs);

  const docsByModel = new Map<string, InferenceDoc[]>();
  for (const doc of docs) {
    const model = normalizeModel(doc.model);
    if (!docsByModel.has(model)) docsByModel.set(model, []);
    docsByModel.get(model)!.push(doc);
  }

  const metricsByModel: Record<string, ModelMetrics> = {};

  for (const [model, modelDocs] of docsByModel) {
    const requests = modelDocs.length;
    const successes = modelDocs.filter((d) => d.ok).length;
    const failures = requests - successes;
    const successDocs = modelDocs.filter((d) => d.ok);

    const ttftValues = collectMetricValues(successDocs, "ttft_ms");
    const providerE2eValues = collectMetricValues(successDocs, "provider_output_tokens_per_second_end_to_end");
    const outputTpsValues = successDocs
      .map((d) => extractOutputTpsPostTtft(d))
      .filter((v): v is number => v !== null);

    const avg = (values: number[]): number | null => {
      if (!values.length) return null;
      return Math.round((values.reduce((a, b) => a + b, 0) / values.length) * 100) / 100;
    };

    metricsByModel[model] = {
      requests,
      successes,
      failures,
      success_rate_percent: requests > 0 ? Math.round((successes / requests) * 100 * 100) / 100 : null,
      avg_ttft_ms: avg(ttftValues),
      avg_output_tps: avg(outputTpsValues),
      p95_ttft_ms: percentile(ttftValues, 0.95) != null ? Math.round(percentile(ttftValues, 0.95)! * 100) / 100 : null,
      avg_provider_tps_end_to_end: avg(providerE2eValues),
    };
  }

  for (const model of ALL_MODELS) {
    if (!metricsByModel[model]) {
      metricsByModel[model] = {
        requests: 0,
        successes: 0,
        failures: 0,
        success_rate_percent: null,
        avg_ttft_ms: null,
        avg_output_tps: null,
        p95_ttft_ms: null,
        avg_provider_tps_end_to_end: null,
      };
    }
  }

  const bucketsByModel = new Map<string, Map<string, BucketData>>();
  for (const point of runTrendPoints) {
    const ts = point.timestamp;
    const model = point.model;

    const bucket = new Date(
      Date.UTC(ts.getUTCFullYear(), ts.getUTCMonth(), ts.getUTCDate(), ts.getUTCHours(), 0, 0, 0),
    );
    const bucketKey = bucket.toISOString();

    if (!bucketsByModel.has(model)) bucketsByModel.set(model, new Map());
    const modelBuckets = bucketsByModel.get(model)!;
    const existing = modelBuckets.get(bucketKey) || initBucketData();

    if (point.output_tps !== undefined && Number.isFinite(point.output_tps)) {
      existing.output_sum += point.output_tps;
      existing.output_count += 1;
    }
    if (point.ttft_ms !== undefined && Number.isFinite(point.ttft_ms)) {
      existing.ttft_sum += point.ttft_ms;
      existing.ttft_count += 1;
    }

    modelBuckets.set(bucketKey, existing);
  }

  const failureBucketsByModel = new Map<string, Set<string>>();
  for (const doc of trendDocs) {
    if (doc.ok) continue;
    const ts = doc.timestamp;
    if (!(ts instanceof Date) || isNaN(ts.getTime())) continue;

    const model = normalizeModel(doc.model);
    const bucket = new Date(
      Date.UTC(ts.getUTCFullYear(), ts.getUTCMonth(), ts.getUTCDate(), ts.getUTCHours(), 0, 0, 0),
    );
    const bucketKey = bucket.toISOString();

    if (!failureBucketsByModel.has(model)) failureBucketsByModel.set(model, new Set());
    const modelFailureBuckets = failureBucketsByModel.get(model)!;
    modelFailureBuckets.add(bucketKey);
  }

  const bucketSizeMs = 60 * 60 * 1000;
  const firstBucketMs = Math.floor(trendWindowStart.getTime() / bucketSizeMs) * bucketSizeMs;
  const lastBucketMs = Math.floor(trendWindowEnd.getTime() / bucketSizeMs) * bucketSizeMs;

  const trendByModel: Record<string, Array<{ timestamp: string; output_tps?: number; ttft_ms?: number }>> = {};
  const failureByModel: Record<string, Array<{ timestamp: string }>> = {};

  for (const [model, modelBuckets] of bucketsByModel) {
    const modelTrend: Array<{ timestamp: string; output_tps?: number; ttft_ms?: number }> = [];

    let bucketCursor = new Date(firstBucketMs);
    while (bucketCursor.getTime() <= lastBucketMs) {
      const bucketKey = bucketCursor.toISOString();
      const data = modelBuckets.get(bucketKey);

      modelTrend.push({
        timestamp: bucketKey,
        output_tps: data && data.output_count > 0 ? Math.round((data.output_sum / data.output_count) * 1000) / 1000 : undefined,
        ttft_ms: data && data.ttft_count > 0 ? Math.round((data.ttft_sum / data.ttft_count) * 100) / 100 : undefined,
      });

      bucketCursor = new Date(bucketCursor.getTime() + bucketSizeMs);
    }

    trendByModel[model] = modelTrend;

    const modelFailures = failureBucketsByModel.get(model);
    if (modelFailures) {
      failureByModel[model] = Array.from(modelFailures.values())
        .sort((a, b) => a.localeCompare(b))
        .map((timestamp) => ({ timestamp }));
    } else {
      failureByModel[model] = [];
    }
  }

  for (const model of ALL_MODELS) {
    if (!trendByModel[model]) {
      const emptyTrend: Array<{ timestamp: string; output_tps?: number; ttft_ms?: number }> = [];
      let bucketCursor = new Date(firstBucketMs);
      while (bucketCursor.getTime() <= lastBucketMs) {
        emptyTrend.push({ timestamp: bucketCursor.toISOString() });
        bucketCursor = new Date(bucketCursor.getTime() + bucketSizeMs);
      }
      trendByModel[model] = emptyTrend;
    }
    if (!failureByModel[model]) {
      failureByModel[model] = [];
    }
  }

  const errorBreakdown = new Map<string, number>();
  for (const d of failureDocs) {
    const errorType = d.error?.type || "unknown_error";
    errorBreakdown.set(errorType, (errorBreakdown.get(errorType) || 0) + 1);
  }

  const models = [...new Set([...docsByModel.keys(), ...ALL_MODELS])].sort((a, b) => a.localeCompare(b));
  const latestTs = docs[docs.length - 1]?.timestamp || trendDocs[trendDocs.length - 1]?.timestamp;

  return {
    window: { hours: requestedHours, start: toIso(trendWindowStart), end: toIso(trendWindowEnd) },
    metrics_by_model: metricsByModel,
    trend_by_model: trendByModel,
    failure_by_model: failureByModel,
    errors: Array.from(errorBreakdown.entries()).map(([type, count]) => ({ type, count })).sort((a, b) => b.count - a.count),
    models,
    endpoint_family: ENDPOINT_FAMILY_CODING_PLAN,
    latest_document_timestamp: toIso(latestTs instanceof Date ? latestTs : null),
    generated_at: toIso(new Date()),
  };
}
